frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-design/mooclet-policy-editor/mooclet-policy-editor.component.ts

```diff
@@ -0,0 +1,104 @@
+import { ExperimentVM } from '../../../../../../core/experiments/store/experiments.model';
+
+@Component({
+  selector: 'app-mooclet-policy-editor',
+  templateUrl: './mooclet-policy-editor.component.html',
+  styleUrls: ['./mooclet-policy-editor.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class MoocletPolicyEditorComponent implements OnInit {
+  @Input() experimentInfo: ExperimentVM;
+  @Input() experimentName: string;
+  @Input() currentAssignmentAlgorithm: string;
+
+  @ViewChild('policyEditor', { static: false }) policyEditor: JsonEditorComponent;
+
+  options = new JsonEditorOptions();
+  defaultPolicyParametersForAlgorithm: MoocletTSConfigurablePolicyParametersDTO;
+  moocletPolicyParametersErrors$ = new BehaviorSubject<ValidationError[]>([]);
+  editorValue$ = new Subject<any>();
+
+  ngOnInit() {
+    this.setupEditor();
+  }
+
+  setupEditor() {
+    // Only create default parameters if we don't have existing ones
+    if (!this.experimentInfo?.moocletPolicyParameters) {
+      this.defaultPolicyParametersForAlgorithm = new MOOCLET_POLICY_SCHEMA_MAP[this.currentAssignmentAlgorithm]();
+      this.defaultPolicyParametersForAlgorithm.outcome_variable_name = `${this.experimentName
+        .trim()
+        .toUpperCase()
+        .replace(/ /g, '_')}_REWARD_VARIABLE`;
+    } else {
+      // Use existing parameters from backend when editing
+      this.defaultPolicyParametersForAlgorithm = this.experimentInfo.moocletPolicyParameters;
+    }
+
+    this.options.mode = 'code';
+    this.options.statusBar = false;
+
+    // Set up value change listener for the editor value
+    // This feels hacky but it ensures that editor value is always getting updated validation
+    this.options.onChange = () => {
+      try {
+        const value = this.policyEditor.get();
+        this.editorValue$.next(value);
+      } catch {
+        // Invalid JSON in editor (The [x] icon will be displayed in the editor)
+      }
+    };
+
+    // Set up validation pipeline for feedback while typing
+    // TODO: Enforce stricter validation (disallow missing props) and improve errors
+    this.editorValue$
+      .pipe(
+        debounceTime(300),
+        switchMap((jsonValue) => this.validateMoocletPolicyParameters(jsonValue))
+      )
+      .subscribe((errors) => {
+        this.moocletPolicyParametersErrors$.next(errors);
+      });
+  }
+
+  resetPolicyParameters() {
+    this.defaultPolicyParametersForAlgorithm = new MOOCLET_POLICY_SCHEMA_MAP[this.currentAssignmentAlgorithm]();
+    this.defaultPolicyParametersForAlgorithm.outcome_variable_name = `${this.experimentName
+      .trim()
+      .toUpperCase()
+      .replace(/ /g, '_')}_REWARD_VARIABLE`;
+
+    // Force editor to update with new default values
+    if (this.policyEditor) {
+      const jsonValue = JSON.parse(JSON.stringify(this.defaultPolicyParametersForAlgorithm));
```

Comment:
porting over my comment, what's the need for `JSON.parse(JSON.stringify(`?

Reply:
I believe we need deep cloning here since `defaultPolicyParametersForAlgorithm` contains nested objects (`Prior`, `CurrentPosteriors`). The spread operator would only create a shallow copy, leaving nested objects as references. We could use `structuredClone()` instead if we want to modernize this, but `JSON.parse(JSON.stringify())` works fine for our data types.

Reply:
> what's JSON.parse(JSON.stringify up to? that looks like the old js trick of deep cloning, but do we need a clone here? if we do {...obj } syntax is simpler

I believe we need deep cloning here since `defaultPolicyParametersForAlgorithm` contains nested objects (`Prior`, `CurrentPosteriors`). The spread operator would only create a shallow copy, leaving nested objects as references. We could use `structuredClone()` instead if we want to modernize this, but `JSON.parse(JSON.stringify())` works fine for our data types.

Reply:
> what's the need for JSON.parse(JSON.stringify(?

The `JSON.parse/stringify` is needed here because the JSON editor component requires a plain JSON object, not a class instance. It converts our `MoocletTSConfigurablePolicyParametersDTO` instance to a plain object format that the editor can handle.